C++ parser coredump on C++23 if consteval

[best_input.cpp](https://github.com/user-attachments/files/27104511/best_input.cpp)

Hello ESBMC team,

This is Eduardo Valentin. Thank you for the building ESBMC! It's an awesome tool.

I am attempting to use ESBMC to generate harnesses and CEXes for [OpenSMA](https://github.com/NVIDIA/OpenSMA). 

While attempt to generate a CEX for the gpio driver of the [OpenSMA](https://github.com/NVIDIA/OpenSMA/blob/vr_sma/src/nv/gpio/driver.cpp) code base, using ESBMC 8.2, I am seeing a consistent coredump.

See the preprocessed version of the driver.cpp attached [driver.i.cpp](https://github.com/user-attachments/files/27104306/driver.i.cpp).

I was also able to `creduce` the C++ file and reduce the crash dump trigger to the following 
[simple lines](https://github.com/user-attachments/files/27104410/best_input.cpp):

```
$ cat best_input.cpp
namespace {
bool a() {
  if consteval {}
}

```

and this is the esbmc command:

```
$ 


esbmc best_input.cpp \
         --timeout 10s \
         --unwind 20 \
         --function read_virtual_physical_gpio \
         --multi-property \
         --overflow-check \
         --force-malloc-success \
         --memory-leak-check \
         --struct-fields-check \
         --ub-shift-check \
         --unsigned-overflow-check \
         --cex-output read_virtual_physical_gpio \
         --incremental-bmc \
         --std c++23 \
         -DPDK_CMN_LOG_DBG_LEVEL_CONSOLE=0 \
         -DPDK_CMN_LOG_DBG_LEVEL_PERSISTENT=0 \
         -DPDK_CMN_LOG_MAX_MSG_LENGTH=256 \
         -DPDK_CMN_LOG_SL_CONSOLE=1 \
         -DPDK_CMN_LOG_SL_PERSISTENT=1

ESBMC version 8.2.0 64-bit x86_64 linux
Target: 64-bit little-endian x86_64-unknown-linux with esbmclibc
[PROGRESS] Parsing driver.i.cpp
driver.i.cpp:4:1: warning: non-void function does not return a value [-Wreturn-type]
    4 | }
      | ^
[PROGRESS] Converting
Segmentation fault (core dumped)


```

--- a/regression/esbmc-cpp20/cpp/github_4196/main.cpp
+++ b/regression/esbmc-cpp20/cpp/github_4196/main.cpp
@@ -1,0 +1,21 @@
+#include <cassert>
+
+int run()
+{
+  int x = 0;
+  if consteval
+  {
+    x = 42;
+  }
+  else
+  {
+    x = 1;
+  }
+  return x;
+}
+
+int main()
+{
+  assert(run() == 1);
+  return 0;
+}

--- a/regression/esbmc-cpp20/cpp/github_4196/test.desc
+++ b/regression/esbmc-cpp20/cpp/github_4196/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.cpp
+--std c++23 --incremental-bmc
+^VERIFICATION SUCCESSFUL$

--- a/regression/esbmc-cpp20/cpp/github_4196_fail/main.cpp
+++ b/regression/esbmc-cpp20/cpp/github_4196_fail/main.cpp
@@ -1,0 +1,17 @@
+#include <cassert>
+
+int run()
+{
+  int x = 0;
+  if !consteval
+  {
+    x = 7;
+  }
+  return x;
+}
+
+int main()
+{
+  assert(run() == 0);
+  return 0;
+}

--- a/regression/esbmc-cpp20/cpp/github_4196_fail/test.desc
+++ b/regression/esbmc-cpp20/cpp/github_4196_fail/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.cpp
+--std c++23 --incremental-bmc
+^VERIFICATION FAILED$

--- a/regression/esbmc-cpp20/cpp/github_4196_no_else/main.cpp
+++ b/regression/esbmc-cpp20/cpp/github_4196_no_else/main.cpp
@@ -1,0 +1,17 @@
+#include <cassert>
+
+int run()
+{
+  int x = 0;
+  if consteval
+  {
+    x = 42;
+  }
+  return x;
+}
+
+int main()
+{
+  assert(run() == 0);
+  return 0;
+}

--- a/regression/esbmc-cpp20/cpp/github_4196_no_else/test.desc
+++ b/regression/esbmc-cpp20/cpp/github_4196_no_else/test.desc
@@ -1,0 +1,4 @@
+CORE
+main.cpp
+--std c++23 --incremental-bmc
+^VERIFICATION SUCCESSFUL$

--- a/src/clang-c-frontend/clang_c_convert.cpp
+++ b/src/clang-c-frontend/clang_c_convert.cpp
@@ -2521,10 +2521,10 @@ bool clang_c_convertert::get_expr(const clang::Stmt &stmt, exprt &new_expr)
         if (get_expr(*runtime_branch, new_expr))
           return true;
         convert_expression_to_code(new_expr);
-       }
-       else
-         new_expr = code_skipt();
-       break;
+      }
+      else
+        new_expr = code_skipt();
+      break;
     }
 
     const clang::Stmt *cond_expr = ifstmt.getConditionVariableDeclStmt();

--- a/src/clang-c-frontend/clang_c_convert.cpp
+++ b/src/clang-c-frontend/clang_c_convert.cpp
@@ -2513,9 +2513,8 @@ bool clang_c_convertert::get_expr(const clang::Stmt &stmt, exprt &new_expr)
     // runtime-active branch only (or skip entirely).
     if (ifstmt.isConsteval())
     {
-      const clang::Stmt *runtime_branch = ifstmt.isNegatedConsteval()
-                                            ? ifstmt.getThen()
-                                            : ifstmt.getElse();
+      const clang::Stmt *runtime_branch =
+        ifstmt.isNegatedConsteval() ? ifstmt.getThen() : ifstmt.getElse();
 
       if (runtime_branch == nullptr)
       {

--- a/src/clang-c-frontend/clang_c_convert.cpp
+++ b/src/clang-c-frontend/clang_c_convert.cpp
@@ -2516,19 +2516,15 @@ bool clang_c_convertert::get_expr(const clang::Stmt &stmt, exprt &new_expr)
       const clang::Stmt *runtime_branch =
         ifstmt.isNegatedConsteval() ? ifstmt.getThen() : ifstmt.getElse();
 
-      if (runtime_branch == nullptr)
+      if (runtime_branch != nullptr)
       {
-        new_expr = code_skipt();
-        break;
-      }
-
-      exprt branch_expr;
-      if (get_expr(*runtime_branch, branch_expr))
-        return true;
-
-      convert_expression_to_code(branch_expr);
-      new_expr = branch_expr;
-      break;
+        if (get_expr(*runtime_branch, new_expr))
+          return true;
+        convert_expression_to_code(new_expr);
+       }
+       else
+         new_expr = code_skipt();
+       break;
     }
 
     const clang::Stmt *cond_expr = ifstmt.getConditionVariableDeclStmt();

--- a/src/clang-c-frontend/clang_c_convert.cpp
+++ b/src/clang-c-frontend/clang_c_convert.cpp
@@ -2508,6 +2508,30 @@ bool clang_c_convertert::get_expr(const clang::Stmt &stmt, exprt &new_expr)
   {
     const clang::IfStmt &ifstmt = static_cast<const clang::IfStmt &>(stmt);
 
+    // C++23 `if consteval` / `if !consteval` carries no condition in the AST.
+    // At runtime the consteval branch is never taken, so lower to the
+    // runtime-active branch only (or skip entirely).
+    if (ifstmt.isConsteval())
+    {
+      const clang::Stmt *runtime_branch = ifstmt.isNegatedConsteval()
+                                            ? ifstmt.getThen()
+                                            : ifstmt.getElse();
+
+      if (runtime_branch == nullptr)
+      {
+        new_expr = code_skipt();
+        break;
+      }
+
+      exprt branch_expr;
+      if (get_expr(*runtime_branch, branch_expr))
+        return true;
+
+      convert_expression_to_code(branch_expr);
+      new_expr = branch_expr;
+      break;
+    }
+
     const clang::Stmt *cond_expr = ifstmt.getConditionVariableDeclStmt();
     if (cond_expr == nullptr)
       cond_expr = ifstmt.getCond();

